executor/aggregate.go

```diff
@@ -949,6 +958,55 @@ func (e *HashAggExec) getPartialResults(groupKey string) []aggfuncs.PartialResul
 	return partialResults
 }
 
+func (e *HashAggExec) executeAllFirstRow(ctx context.Context, chk *chunk.Chunk) (err error) {
+	for {
+		if e.firstRowProcessed == e.childResult.NumRows() {
+			mSize := e.childResult.MemoryUsage()
+			e.firstRowProcessed = 0
+			err := Next(ctx, e.children[0], e.childResult)
+			e.memTracker.Consume(e.childResult.MemoryUsage() - mSize)
+			if err != nil {
+				return err
+			}
+			// no more data.
+			if e.childResult.NumRows() == 0 {
+				return nil
+			}
+			e.groupKeyBuffer, err = getGroupKey(e.ctx, e.childResult, e.groupKeyBuffer, e.GroupByItems)
+			if err != nil {
+				return err
+			}
+		}
+
+		failpoint.Inject("unparallelHashAggError", func(val failpoint.Value) {
+			if val.(bool) {
+				failpoint.Return(errors.New("HashAggExec.unparallelExec error"))
+			}
+		})
+
+		for e.firstRowProcessed < e.childResult.NumRows() {
+			groupKey := string(e.groupKeyBuffer[e.firstRowProcessed]) // do memory copy here, because e.groupKeyBuffer may be reused.
```

Comment:
It seems that `groupKey` is only used in `e.groupSet.Exist` and `e.groupSet.Insert`.
Can we use `hack.String` instead of `string` here?

Reply:
I think yes.